### PR TITLE
(PUP-9994) implement $trusted['external'] facts

### PR DESCRIPTION
**Pull request by tphoney**

### CODE REVIEW

**Comment by tphoney** [comment]
Below is the output of the test running through
```
    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) 15:05:50$ cat /etc/puppetlabs/puppet/puppet.conf
      [main]
      server = brazen-clot.delivery.puppetlabs.net
      trusted_external_command = /tmp/trusted_external_facts.M8O9Jh/fact.sh

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) executed in 0.51 seconds

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) 15:05:51$ puppet apply -e "notify {'trusted facts': message => $::trusted }"
      Notice: Compiled catalog for moot-static.delivery.puppetlabs.net in environment production in 0.01 seconds
      Notice: {"authenticated"=>"local", "certname"=>"moot-static.delivery.puppetlabs.net", "extensions"=>{}, "hostname"=>"moot-static", "domain"=>"delivery.puppetlabs.net", "external"=>{"doot"=>"moot-static.delivery.puppetlabs.net"}}
      Notice: /Stage[main]/Main/Notify[trusted facts]/message: defined 'message' as {
        'authenticated' => 'local',
        'certname' => 'moot-static.delivery.puppetlabs.net',
        'extensions' => {

        },
        'hostname' => 'moot-static',
        'domain' => 'delivery.puppetlabs.net',
        'external' => {
          'doot' => 'moot-static.delivery.puppetlabs.net'
        }
      }
      Notice: Applied catalog in 0.01 seconds

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) executed in 2.40 seconds

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) 15:05:53$ puppet config print config --section master
      /etc/puppetlabs/puppet/puppet.conf

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) executed in 1.33 seconds

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) 15:05:55$ if [ -f '/tmp/trusted_external_facts.M8O9Jh/puppet.conf.bak' ]; then cat '/tmp/trusted_external_facts.M8O9Jh/puppet.conf.bak' > '/etc/puppetlabs/puppet/puppet.conf'; rm -f '/tmp/trusted_external_facts.M8O9Jh/puppet.conf.bak'; fi

    moot-static.delivery.puppetlabs.net (moot-static.delivery.puppetlabs.net) executed in 0.51 seconds
Begin teardown
End teardown
tests/ssl/trusted_external_facts.rb passed in 20.15 seconds
      Test Suite: tests @ 2019-11-08 15:05:35 +0000

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 20.15 seconds
      Average Test Time: 20.15 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:

No tests to run for suite 'post_suite'
```

**Comment by tphoney** [comment]
this supersedes #7694

File: acceptance/tests/ssl/trusted_external_facts.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        }
+      }
+      backup_file = backup_the_file(agent, puppet_config(agent, 'confdir', section: 'master'), testdirs[agent], 'puppet.conf')
+      lay_down_new_puppet_conf agent, puppet_conf, testdirs[agent]
+
+      on(agent, puppet("apply", "-e", "\"notify {'trusted facts': message => $::trusted }\""), :acceptable_exit_codes => [0,2])
+      assert_match(/external/, stdout, "Trusted fact hash contains external key")
+      assert_match(/doot.*#{agent}/, stdout, "trusted facts contains certname")
+
+      restore_puppet_conf_from_backup(agent, backup_file)
```
Comment (joshcooper) [inline acceptance/tests/ssl/trusted_external_facts.rb]: If the `assert_match` fails, then we never restore `puppet.conf` and the agent will be left in a bad state. You could add an ensure block, but in cases like this, just pass the script as a command line arg:

```ruby
on(agent, puppet("apply", "--trusted_external_command", external_trusted_fact_script_path, ...)
```

File: lib/puppet/defaults.rb [hunk 2]
```diff
@@ -306,7 +306,6 @@ def self.default_vendormoduledir
         :default  => ! Puppet::Util::Platform.windows?,
         :type     => :boolean,
         :desc     => "Whether Puppet should manage the owner, group, and mode of files it uses internally.
-
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: nit unrelated change

File: lib/puppet/context/trusted_information.rb [hunk 3]
```diff
@@ -51,17 +61,42 @@ def self.remote(authenticated, node_name, certificate)
...
+      object.each {|k, v| deep_freeze(k); deep_freeze(v) }
+      object.freeze
+    when NilClass, Numeric, TrueClass, FalseClass
+      # do nothing
+    when String
+      object.freeze
+    else
+      raise Puppet::Error, _("Unsupported data type: '%{klass}'") % { klass: object.class }
+    end
+    object
```
Comment (joshcooper) [inline lib/puppet/context/trusted_information.rb]: `retrieve_trusted_external` and `deep_freeze` should be private class methods

File: lib/puppet/context/trusted_information.rb [hunk 4]
```diff
@@ -51,17 +61,42 @@ def self.remote(authenticated, node_name, certificate)
           [ext['oid'].freeze, ext['value'].freeze]
         end]
       end
-      new('remote', node_name, extensions)
+      new('remote', node_name, extensions, external)
     else
-      new(false, nil, {})
+      new(false, nil, {}, external)
```
Comment (joshcooper) [inline lib/puppet/context/trusted_information.rb]: If `trusted_external_command` is not defined, then `retrieve_trusted_external` will return nil, which is passed in here and above on line 64. As a result `Puppet::Context::TrustedInformation#external` may return nil or a Hash. I'm thinking it would be better if `retreive_trusted_external` always returned a `Hash`

**Comment by joshcooper** [comment]
jenkins please test this on redhat7-64a,windows2019-64a

File: acceptance/tests/ssl/trusted_external_facts.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+      puppet_conf = {
+        'main' => {
+          'trusted_external_command' => external_trusted_fact_script_path
+        }
+      }
+      backup_file = backup_the_file(agent, puppet_config(agent, 'confdir', section: 'master'), testdirs[agent], 'puppet.conf')
+      lay_down_new_puppet_conf agent, puppet_conf, testdirs[agent]
+
+      on(agent, puppet("apply", "-e", "\"notify {'trusted facts': message => $::trusted }\""), :acceptable_exit_codes => [0,2])
```
Comment (joshcooper) [inline acceptance/tests/ssl/trusted_external_facts.rb]: This test won't work on Windows, because it relies on a bash script. It would be better to create the script on the server to verify that puppetserver can execute the script and add the values to the catalog.

**Comment by joshcooper** [comment]
Superseded by https://github.com/puppetlabs/puppet/pull/7824
